### DISCUSSION TITLE
How extracted timestamps when using a beam search decoder?

### DISCUSSION THREAD

**Discussion by Slava715** [NVIDIA-NeMo/NeMo#2577] [Q&A] [answered]
The notebook https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/Offline_ASR.ipynb has a good example of how to extract the timestamps of words using the QuartzNet model.
Is it possible to do something like this using Conformer-CTC and a language model KenLM?

**Comment by Slava715** [comment, marked as answer]
This was solved by using: https://github.com/parlance/ctcdecode

**Reply by titu1994** [reply]
Thanks! I didn't see this message before. Does it work with Subword models ? Or just QuartzNet?

**Reply by Slava715** [reply]
Yes, it works with subwords, you need to add "_" token to the end of the dictionary and indicate its position:

vocab = asr_model.decoder.vocabulary
vocab.append("_")
TOKEN_OFFSET = 100
beam_search_lm = CTCBeamDecoder(
      [chr(idx + TOKEN_OFFSET) for idx in range(len(vocab))],
      model_path=lm_path,
      alpha=1.0,
      beta=1.0,
      beam_width=128,
      num_processes=max(os.cpu_count(), 1),
      blank_id=128,
      log_probs_input=False)

**Reply by dbkalaria** [reply]
Hi @Slava715,  have you manage to get the timestamp from conformer-CTC model? If yes can you help me in getting the timestamp. I've tried the step you mentioned above but getting some random result. It would be great if you help me.

**Reply by Slava715** [reply]
Try this example:

`import nemo
import nemo.collections.asr as nemo_asr
from nemo.core.classes import IterableDataset
from nemo.core.neural_types import NeuralType, AudioSignal, LengthsType

import torch
from torch.utils.data import DataLoader

from ctcdecode import CTCBeamDecoder

import numpy as np

import os, time

import math


# sample rate, Hz
SAMPLE_RATE = 16000

# duration of signal frame, seconds
FRAME_LEN = 40.0

# number of audio channels (expect mono signal)
CHANNELS = 1

CHUNK_SIZE = int(FRAME_LEN*SAMPLE_RATE)

TOKEN_OFFSET = 100

TIME_STEP = 0.04

TIME_PAD = 1

# simple data layer to pass audio signal
class AudioDataLayer(IterableDataset):
    @property
    def output_types(self):
        return {
            'audio_signal': NeuralType(('B', 'T'), AudioSignal(freq=self._sample_rate)),
            'a_sig_length': NeuralType(tuple('B'), LengthsType()),
        }

    def __init__(self, sample_rate):
        super().__init__()
        self._sample_rate = sample_rate
        self.output = True
        
    def __iter__(self):
        return self
    
    def __next__(self):
        if not self.output:
            raise StopIteration
        self.output = False
        return self.signal, self.signal_shape 
        
    def set_signal(self, signal):
        self.signal = signal
        self.signal_shape = torch.as_tensor(len(signal), dtype=torch.int64)
        self.output = True

    def __len__(self):
        return 1
    
    
# class for streaming frame-based ASR
# 1) call transcribe(frame) to do ASR on
#    contiguous signal's frames
class FrameASR:
    
    def __init__(self, frame_len=40, model_path="", lm_path=""):
        
        '''
        Args:
          frame_len: frame's duration, seconds
        '''
        
        asr_model = nemo_asr.models.EncDecCTCModel.restore_from(model_path)
        asr_model.eval();
        self.asr_model = asr_model.to(asr_model.device)
        
        self.sr = SAMPLE_RATE
        self.frame_len = frame_len
        self.n_frame_len = int(frame_len * self.sr)
        
        self.data_layer = AudioDataLayer(sample_rate=self.asr_model.cfg.preprocessor.sample_rate)    
        self.data_loader = DataLoader(self.data_layer, batch_size=1, collate_fn=self.data_layer.collate_fn)
        
        self.vocab = self.asr_model.decoder.vocabulary
        self.vocab.append("_")
        
        self.beam_search_lm = CTCBeamDecoder(
            [chr(idx + TOKEN_OFFSET) for idx in range(len(self.vocab))],
            model_path=lm_path,
            alpha=0.75,
            beta=1.0,
            beam_width=128,
            num_processes=max(os.cpu_count(), 1),
            blank_id=128,
            cutoff_prob=1.0,
            cutoff_top_n=40,
            log_probs_input=False)
        
    def _decode(self, frame, chunk_pad):
        logits = self.infer_signal(self.asr_model, frame).cpu().numpy()[0]
        # print(logits.shape)
        decoded = self._beam_decoder(logits, chunk_pad)
        return decoded
    
    @torch.no_grad()
    def transcribe(self, frame=None, chunk_pad = 0):
        if frame is None:
            frame = np.zeros(shape=self.n_frame_len, dtype=np.float32)
        return self._decode(frame, chunk_pad)
        
    def _beam_decoder(self, logits, chunk_pad):
        probs = softmax(logits)
        probs_seq = torch.FloatTensor(probs)
        
        beam_results, beam_scores, timesteps, out_lens = self.beam_search_lm.decode(probs_seq.unsqueeze(0))
        
        beam_res = beam_results[0][0][:out_lens[0][0]].cpu().numpy()
        times = timesteps[0][0][:out_lens[0][0]].cpu().numpy()
        lens = out_lens[0][0]
        
        transcript = self.asr_model.tokenizer.ids_to_text(beam_res)
        
        if len(times) > 0:
            if times[0] < TIME_PAD:
                start = chunk_pad
                end = (times[0] + TIME_PAD*2) * TIME_STEP + chunk_pad
            else:
                start = (times[0] - TIME_PAD) * TIME_STEP + chunk_pad
                end = (times[0] + TIME_PAD) * TIME_STEP + chunk_pad
                
            tocken_prev = self.vocab[int(beam_res[0])]
            word = tocken_prev
            
            result = []
            
            for n in range(1,lens):
                tocken = self.vocab[int(beam_res[n])]
                
                if tocken[0] == "#":
                    word = word + tocken[2:]
                    
                elif tocken[0] == "-" or tocken_prev[0] == "-":
                    word = word + tocken
                    
                else:
                    if start > end:
                        result_word = { 'start': round(end, 3), 'end': round(start, 3), 'word': word}
                    else:
                        result_word = { 'start': round(start, 3), 'end': round(end, 3), 'word': word}
                        
                    result.append(result_word)
                    
                    if times[n] < TIME_PAD:
                        start = chunk_pad
                    else:
                        start = (times[n] - TIME_PAD) * TIME_STEP + chunk_pad

                    word = tocken
                    
                    
                if times[n] < TIME_PAD:
                    end = (times[n] + TIME_PAD*2) * TIME_STEP + chunk_pad
                else:
                    end = (times[n] + TIME_PAD) * TIME_STEP + chunk_pad
                
                tocken_prev = tocken
                
                
            if start > end:
                result_word = { 'start': round(end, 3), 'end': round(start, 3), 'word': word}
            else:
                result_word = { 'start': round(start, 3), 'end': round(end, 3), 'word': word}
                
            result.append(result_word)
            
        else:
            print(transcript)
            result = []
            
        return transcript, result
        
    def infer_signal(self, model, signal):
        self.data_layer.set_signal(signal)
        batch = next(iter(self.data_loader))
        audio_signal, audio_signal_len = batch
        audio_signal, audio_signal_len = audio_signal.to(self.asr_model.device), audio_signal_len.to(self.asr_model.device)
        log_probs, encoded_len, predictions = model.forward(
            input_signal=audio_signal, input_signal_length=audio_signal_len
        )
        return log_probs
    
    
def softmax(x):
    e = np.exp(x - np.max(x))
    return e / e.sum(axis=-1).reshape([x.shape[0], 1])

from pydub import AudioSegment

MODEL_PATH="disk/vol1/asr_scripts/ctc_large-it4-1.nemo"
LM_PATH="disk/vol1/asr_scripts/sum_corp_conf.bin"

asr = FrameASR(frame_len=FRAME_LEN, model_path=MODEL_PATH, lm_path=LM_PATH)

def asr_process(file_name):
    
    file_bin_data = AudioSegment.from_file(file_name, "wav").set_frame_rate(16000).get_array_of_samples()
    
    file_np = np.frombuffer(file_bin_data, dtype=np.int16)/32768.
    
    file_tensor = torch.FloatTensor(file_np)
    
    sum_text = ""
    sum_res = []
    
    if len(file_tensor) > CHUNK_SIZE:
        get_chunks = torch.tensor_split(file_tensor, math.ceil(len(file_tensor)/CHUNK_SIZE))
        chunk_pad = 0
        for chunk in get_chunks:
            text, res = asr.transcribe(frame = chunk, chunk_pad = chunk_pad/SAMPLE_RATE)
            chunk_pad = chunk_pad + len(chunk)
            sum_text = sum_text + " " + text
            sum_res = sum_res + res
            
    elif len(file_tensor) > 0:
        text, res = asr.transcribe(frame = file_tensor, chunk_pad = 0)
        sum_text = sum_text + " " + text
        sum_res = sum_res + res
        
    sum_text = sum_text[1:].replace("  ", " ").replace('#', '')
    return sum_text, sum_res`

**Reply by dbkalaria** [reply]
@Slava715 can you tell me which language model have you used?

**Reply by Slava715** [reply]
https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/asr_language_modeling.html#ngram-modeling

**Reply by shibanaz** [reply]
@Slava715 :

What is LM_PATH and MODEL_PATH here

**Reply by drorpa** [reply]
I'm trying to use https://github.com/parlance/ctcdecode to get the timestamps, but i receive an error:

> [ctcdecode/src/ctc_beam_search_decoder.cpp:62] FATAL: "(probs_seq[i].size()) == (vocabulary.size())" check failed. The shape of probs_seq does not match with the shape of the vocabulary

when i look at the variables, i see that the vocab len is 28, but the probs have a dimension of size 29
any advice how to fix it?


This is my code:
```
asr_model = nemo_asr.models.EncDecCTCModel.from_pretrained(model_name='stt_en_quartznet15x5', strict=False)
labels=list(asr_model.decoder.vocabulary)

logits = asr_model.transcribe([path], logprobs=True)[0]
outputs = torch.from_numpy(np.expand_dims(logits, axis=0))

lm_path = 'lowercase_3-gram.pruned.1e-7.arpa'

decoder = CTCBeamDecoder(
    labels=labels,
    model_path=lm_path,
    alpha=2,
    beta=1.5,
    cutoff_top_n=40,
    cutoff_prob=1.0,
    beam_width=16,
    num_processes=max(os.cpu_count(), 1),
    blank_id=0,
    log_probs_input=False
)

beam_results, beam_scores, timesteps, out_lens = decoder.decode(outputs)
```

The sizes are:
> \> len(labels)
28
\> outputs.shape
torch.Size([1, 7749, 29])

**Reply by titu1994** [reply]
The last index in vocabulary is for CTC blank. Your blank I'd is therefore wrong - it should be 28 rather than 0.

**Reply by drorpa** [reply]
Thanks, does it mean i need to add CTC black to the vocab?

**Reply by titu1994** [reply]
No, look at the arguments of the class you are calling. Blank_id is wrong.

**Reply by drorpa** [reply]
Changing only the blank_id doesn't address the size mismatch, the same error repeats

**Reply by lmessinger** [reply]
+1
